src/secret.rs

```diff
@@ -1,90 +1,31 @@
 //! Utilities for working with secret values. This module includes functionality for overwriting
 //! memory with zeros.
 
-use std::mem::{size_of, size_of_val};
-use std::ops::{Deref, DerefMut};
+use zeroize::Zeroize;
 
-use memsec::memzero;
-
-use crate::Fr;
-
-pub(crate) const FR_SIZE: usize = size_of::<Fr>();
+use crate::{Fr, FrRepr};
 
 /// Overwrites a single field element with zeros.
-pub(crate) fn clear_fr(fr_ptr: *const Fr) {
-    unsafe { memzero(fr_ptr as *mut u8, FR_SIZE) };
-}
-
-pub(crate) struct MemRange {
-    pub ptr: *mut u8,
-    pub n_bytes: usize,
-}
-
-/// Marks a type as containing some secret value.
-pub(crate) trait ContainsSecret {
-    /// Returns the range of memory marked as secret.
-    fn secret_memory(&self) -> MemRange;
-
-    /// Overwrites the secret region of memory with zeros.
-    ///
-    /// This method should be called upon destruction of every type that implements `ContainsSecret`.
-    fn zero_secret(&self) {
-        let MemRange { ptr, n_bytes } = self.secret_memory();
-        unsafe { memzero(ptr, n_bytes) };
-    }
+pub(crate) fn clear_fr(fr: &mut Fr) {
+    let fr_repr_ptr = unsafe { &mut *(fr as *mut Fr as *mut FrRepr) };
+    fr_repr_ptr.0.zeroize();
```

Comment:
```
error[E0605]: non-primitive cast: `&mut &mut pairing::bls12_381::fr::Fr` as `&mut [u64; 4]`
  --> src/secret.rs:10:29
   |
10 |     let fr_bytes = unsafe { &mut fr as &mut [u64; 4] };
   |                             ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: an `as` expression can only be used to convert between primitive types. Consider using the `From` trait
```

Reply:
To completely avoid playing with pointer casting, it's better to implement `Zeroize` for `Fr` in pairing library

Reply:
Maybe we'd have to use `mem::transmute::<&mut Fr, &mut [u8; FR_SIZE]>`? :grimacing:

Reply:
Well, it'll potentially introduce undefined behaviours....

Reply:
```rust
pub(crate) fn clear_fr(fr: &mut Fr) {
    let fr_bytes = unsafe { &mut *(fr as *mut Fr as *mut [u8; 32]) };
    fr_bytes.zeroize();
}
```
You can use this, but I don't see too much difference from above

Reply:
Another option would be to essentially duplicate the implementation [as in `zeroize`](https://github.com/iqlusioninc/crates/blob/a8fe97af203ddcc65d24faf08c1befa12f2b66ea/zeroize/src/lib.rs#L254), but using `Fr::zero` instead of `default`.

Reply:
Why would it introduce more undefined behavior than the other one? Both are `unsafe` and do the same thing, don't they?

Reply:
| transmute is incredibly unsafe. There are a vast number of ways to cause undefined behavior with this function. transmute should be the absolute last resort.

https://doc.rust-lang.org/std/mem/fn.transmute.html

Reply:
OK, let's use this for now then.
```rust
pub(crate) fn clear_fr(fr: &mut Fr) {
    let fr_bytes = unsafe { &mut *(fr as *mut Fr as *mut [u8; sizeof::<Fr>()]) };
    fr_bytes.zeroize();
}
```
I'll look into it on the weekend, and maybe just duplicate the code from `zeroize`.

Reply:
Never mind, let me just merge it as it is.
I might modify it on the weekend.